Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java

```diff
@@ -495,4 +505,12 @@ public int compareTo(OutputDoublePair<T> o) {
         }
     }
 
+    /**
+     * Used to allow FJPs to work with OpenSearch's SecureSM.
+     */
+    private static final class CustomForkJoinWorkerThreadFactory implements ForkJoinPool.ForkJoinWorkerThreadFactory {
```

Comment:
See same method defined in both `KMeansTrainer` and here, not sure if can define in some common place then these 2 classes can reuse directly. So we can just change one place if we need to tune this method in future.

Reply:
I agree it's not a good idea to have duplicated code, but I don't want to have the `CustomForkJoinWorkerThreadFactory` be a public class that becomes part of Tribuo's API. As Tribuo isn't using Java modules to guard implementation packages yet I think the options are make the class public or duplicate the code. If we end up requiring this in more than two places then we can spend some time and design a proper entry point (maybe a method on `org.tribuo.util.Util` which constructs a `ForkJoinPool` with the required parallelism).

Reply:
Yeah, looks a factory/util class generates `ForkJoinPool` is one option, so we can avoid duplicate `CustomForkJoinWorkerThreadFactory` and `System.getSecurityManager() == null ? new ForkJoinPool(numThreads) : new ForkJoinPool(numThreads, THREAD_FACTORY, null, false);`

Reply:
For the moment I'd like to leave it as is. It's all private so we can always change it later to point at a public endpoint when we have more pieces of code which need this functionality.

Reply:
Sure, sound good to me